text/delegatable_cdd/README.md

```diff
@@ -0,0 +1,514 @@
+                    vec![HashSet::<usize>::new(); i],
+                    &precomputed_setup_params,
+                    &precomp_sig_vals,
+                )?
+    ``` 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Most of the implemented logic follows the paper with the corrections called out
+in the code.  
+At several places, computations like e(a, b)<sup>c</sup> have been replaced with
+e(a<sup>c</sup>, b) since they are about 30% faster. Attempts are made to
+replace computations in G2 with computations in G1 inside pairings.  
+Regarding the implementation of `verify_fast` function, it applies this
+observation to pairings: if it needs to be checked that a == b and c == d and e
+== f, then choose a random number `r` and check whether (a-b) + (c-d)\*r +
+(e-f)\*r<sup>2</sup> == 0. This is because this degree 2 polynomial can have at
+most 2 roots and if r is not a root then all the coefficients must be 0. If r is
+a randomly chosen element from a sufficiently large set, then the chances of r
+being a root are negligible.   
+In a pairing scenario if verifier had to check if e(a,b) = 1, e(c, d) = 1 and e(f, g) = 1, 
+    - pick a random value r in Z<sub>p\*</sub> and 
+    - check e(a,b) \* e(c,d)<sup>r</sup> \* e(f,g)<sup>r<sup>2</sup></sup> equals 1 - e(a,b) \* e(c,d)<sup>r</sup> \* e(f,g)<sup>r<sup>2</sup></sup> = e(a,b) \*
+    e(c<sup>r</sup>, d) \* e(f<sup>r<sup>2</sup></sup>, g). Exponent moved to 1st element of pairing since computation in group G1 is cheaper. 
+    - Now use a single multi-pairing rather than 3 pairings to compute e(a,b) \*
+    e(c<sup>r</sup>, d) \* e(f<sup>r<sup>2</sup></sup>, g)
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This solution should not be used when privacy among issuers is desired. Secondly,
+since each issuer can add any attributes in its issued credential, if policies
+for issuance are not properly defined, malicious issuers might be able to issue
+credentials with unintended attributes (by the delegator) leading to privilege
+escalation. 
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+An expensive alternative to delegatable credentials is the holder to get
+credential directly from the root issuer. The expensiveness of this is not just
+computational but operational too.  
+There are some alternatives which offer privacy among issuers as well but have
+different tradeoffs. [Delegatable Attribute-based Anonymous Credentials from
+Dynamically Malleable Signatures](https://eprint.iacr.org/2018/340) requires
+the presence of a trusted third party. [Delegatable Anonymous Credentials from
+Mercurial Signatures](https://eprint.iacr.org/2018/923) does not support
+credentials with attributes.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+Delegatable anonymous credentials have been explored since the last decade and
+the first efficient (somewhat) came in 2009 by Belenkiy et al. in "Randomizable
```

Comment:
Can you provide a hyperlink to Belenkiy?

Reply:
`If you are quoting, you should put quotes around the sentence and give a hyperlink to your source.
`
I will put quotes. I have a link in the footnote form but I now realize that github does not display it nicely like my editor. Will change.